`{CMS,PKCS7}_verify()`: use `certs` parameter (`-certfile` app option) also for chain building

DDvO opened the pull request:
The documentation said in `openssl-smime.pod.in` in the section on the `-certfile` option:
> The supplied certificates can still be used as untrusted CAs however.

which was misleading. It turned out (see the [discussion below
(https://github.com/openssl/openssl/pull/18916#discussion_r1341990659) which meanwhile is resolved)
that this was referring only to the certificates included in the signed message.

~Moreover, the option name was pretty unclear if not misleading.
So changed this to `-untrusted` while retaining the old name as alias for backward compatibility.~

This PR meanwhile has been split into clarifications, which are now done in #22285, and an extension done here:
Use the certificates provided with the `-certfile` option not only as candidate signer certs (as done so far already)
but also as candidate intermediate CA certs for chain building during verification.

This extension also motivated minor improvements (in a separate commit) of `CMS_get1_{certs,crls}()`: make sure they return NULL only on error.

~Moreover, this PR includes (in a further separate commit) some app code refactoring of `smime.c` and `cms.c` to respect the coding guidelines.~

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

t8m commented:
I could approve it for master and potentially 3.1. IMO this should not go into 3.0 and definitely not 1.1.1.

DDvO commented:
> I could approve it for master and potentially 3.1. IMO this should not go into 3.0 and definitely not 1.1.1.

Yeah, meanwhile 1.1.1. is in security maintenance mode only, so removed that label here.
This PR is a bug fix for the `-certfile` option, which so far does not work as documented.
Why do you think it should not go into 3.0 ?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 309 days ago

paulidale reviewed (approved):
Ping @openssl/committers

beldmit reviewed (approved):
LGTM

beldmit commented:
Not sure if we want to merge it to all three branches

paulidale commented:
Hold to answer @beldmit's question about branches.

vdukhovni commented on doc/man1/openssl-smime.pod.in:
```diff
@@ -177,7 +178,7 @@ If not specified triple DES is used. Only used with B<-encrypt>.
 When verifying a message normally certificates (if any) included in
 the message are searched for the signing certificate. With this option
-only the certificates specified in the B<-certfile> option are used.
+only the certificates specified in the B<-untrusted> option are used.
```
This paragraph is IMNSHO very confusing both **before** and **after** the proposed change. Poor grammar aside (what is "verifying a message normally certificates", where does the comma go?):

The text seems to want to say that

1. The immediate (EE) signer of the message is normally obtained from the certificates included in the message.
2. With the new option the EE certificate is instead obtained from the (till now) `-certfile` file.
3. It then seems to say that once the EE certificate is found, some set of certificates is still used as untrusted certificates to build the chain, but rather unclear which!

The key issue is the muddy meaning of *supplied* in "3". Is that the certificates in the message, or the the `-certfile`?  What was the original behaviour?

Finally, I think it is confusing to call this `-untrusted`, when signing. In that context `-certfile` is more clear.

My suggestion is to NOT adopt this PR, until we first clarify the documentation with no other changes, and then consider whether a new option name is warranted.

vdukhovni replied:
Before this PR, it seems that using `-certfile` always ignored any certificates in the message, using only `-certfile` for both the EE cert and chain construction. This could potentially work around outdated issuers in the message.

Now the message has the union of the original and `-certfile` certificates. Is this behaviour change deliberate and desirable? Now also, the EE certificate (again possibly expired in the message) will be taken from the message, rather than `-certfile`.

DDvO replied:
To me, the intended meaning of "The supplied certificates can still be used as untrusted CAs however." is clear: the "supplied certificates" refer to those given with the CLI option (not those in the message).

I agree that the grammar should be improved by adding a comma in the first sentence, 
and the last sentence of this section could be re-phrased to avoid misunderstandings.

This PR so far does not intend to improve the documentation, but this could be added.

DDvO replied:
Unfortunately, the function `PKCS7_add_certificate()` used here is not documented.
I find it even more unfortunate and confusing that the certs that can be added this way play a double role:
they are used for both signer (EE) certs and for untrusted (intermediate) CA certs.

This PR focused on the 'untrusted' aspect, but as @vdukhovni pointed out above, there is also the 'signer' aspect.

I would have been much better if the two aspects had been kept apart, and then it would be straightforward to have both a  `-signer_certs` and an `-untrusted` option rather than the ambivalent `-certfile` option.

I meanwhile agree with Viktor that, given that weird semantics, the old name `-certfile` was better.

DDvO replied:
So I'm going to revert adding the alias `-untrusted`.
I will also add improvements to the documentation.

vdukhovni replied:
But the P.R. also introduced a change in behaviour. Was that intentional?

Previously, `-certfile` *replaced* any certificates included in the message, now it augments them...

DDvO replied:
> Before this PR, it seems that using -certfile always ignored any certificates in the message, using only -certfile for both the EE cert and chain construction.

@vdukhovni, what you describe here is not the effect of using `-certfile` but of using `-nointern`.
Note the context of the documentation paragraph you are commenting here.

DDvO replied:
After having a close look at the sources and comparing also with CMS, I came to the conclusion that the interpretation I gave yesterday:

> To me, the intended meaning of "The supplied certificates can still be used as untrusted CAs however." is clear: the "supplied certificates" refer to those given with the CLI option (not those in the message).

was not the (originally) intended one. Instead, I am meanwhile sure that the "supplied certificates" were meant to denote the ones included in the message. I will clarify this in the documentation.

DDvO replied:
This clarification is meanwhile done in PR #22285.

DDvO replied:
Moreover, the `-untrusted` option (alias) originally added by this PR is gone after the clarifications achieved meanwhile.
So I believe this conversation thread has been resolved.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

vdukhovni commented on apps/smime.c:
```diff
@@ -607,13 +609,17 @@ int smime_main(int argc, char **argv)
+        int i;
+
+        for (i = 0; i < sk_X509_num(untrusted_certs); i++)
+            PKCS7_add_certificate(p7, sk_X509_value(untrusted_certs, i));
```
What is the point of adding these to the `p7` object? It is not clear what's wrong with the original line `610`. In both cases `NOINTERN` will use only the `-certfile` certs, and otherwise both what's in the message and the additional chain certs.

DDvO replied:
The idea was that the certs given with `-certfile` should be made available for chain building during verification.
Yet as you pointed out, doing it this way has an unwanted side effect on the set of certs that are candidates for signer certs.

Yesterday I did not mention that I was already planning to replace this problematic part,
namely by new code in `PKCS7_verify()` that instead combines on-the-fly the certs given with `-certfile` with the certs contained in the message for for use with chain building.
I just didn't have the time to do so, but likely very soon...

vdukhovni commented:
@beldmit, @paulidale This PR did not get quite enough scrutiny during the initial reviews. Please exercise more care in reviewing/approving unfamiliar code. I wasn't familiar with this code either, but the code change in the PR looked odd in combination with what appeared to be just a documentation and CLI flag change that did not suggest there should be any functional difference, and the documentation language was not entirely clear... :-(

t8m commented:
**OTC: This is not a bug fix because it is adding a new option. Not applicable to stable branches currently.**

DDvO commented:
After the feedback by @vdukhovni, I've revised this PR entirely.
Now I've implemented the additional use of the certificates provided with the `-certfile` option as untrusted certs for verification
in a different way that does not interfere with certificates to be placed in the PKCS#7 structure.
This is now clearly a new feature, and the documentation is extended accordingly.
Due to the unfortunate overlap with CMS, I had to repeat this there.

On this occasion I also found and fixed a related issue with `CMS_get1_{certs,crls}()`: make sure they return NULL only on error, which improves their error behavior in a way that IMO is backward compatible.
I also improved the adherence of `apps/{cms.c,smime.c}` to the coding guidelines.

I found and fixed several more parts of the PKCS#7 and CMS documentation that were unclear and incomplete;
since those should also be backported, I moved them to a separate PR: #22285.

DDvO commented:
Hmm, for some weird reason, I cannot reproduce locally the memory leaks reported by the three sanitizer CI runs (using libasan), neither on Linux nor on MacOS.

DDvO commented:
Got rid of the memory leaks and moved app code refactoring to #22304.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 278 days ago

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the master branch. Thank you for your contribution.

job commented:
> On this occasion I also found and fixed a related issue with `CMS_get1_{certs,crls}()`: make sure they return NULL only on error, which improves their error behavior in a way that IMO is backward compatible.

Sadly this change was not backwards compatible. The following code was compliant with documentation and broke starting with OpenSSL 3.4.0.

```
CMS_ContentInfo *cms;
STACK_OF(X509_CRL) *crls;

cms = d2i_CMS_ContentInfo(NULL, &der, len);
crls = CMS_get1_crls(cms);

if (crls != NULL) {
    printf("Signed Object violates RFC 6488: CMS contained CRLs!");
    goto err;
}
```

t8m commented:
Hmm... that's bad, we should probably revert that change.

DDvO commented:
> ```
> crls = CMS_get1_crls(cms);
> 
> if (crls != NULL) {
>     printf("Signed Object violates RFC 6488: CMS contained CRLs!");
>     goto err;
> }
> ```

Oh, right, in this case there is a change of documented behavior - sorry for that!

> Hmm... that's bad, we should probably revert that change.

Maybe we have to. 
Yet then we are back to the problem that if this function (and likewise `CMS_get1_certs(`) returns NULL,
how should the caller know whether this is due to an error or because the list of CRLs (or certs) is empty?
In the new implementation, the latter case is indicated straightforwardly by an empty list being returned.

So the above code would better be adapted as follows, 
making it correct also for the case of errors thrown by `CMS_get1_crls(cms)` :
```
if (sk_X509_CRL_num(crls) > 0) {
    printf("Signed Object violates RFC 6488: CMS contained CRLs!");
    goto err;
}
```

job commented:
Also of note, `CMS_get0_signers()` returns `NULL` on zero signers, not an empty `STACK` - so this change to some degree further distorts internal consistency.

DDvO commented:
> `errno` checking can be used to figure out if there was an allocation failure when calling `CMS_get1_crls()`, and a-priori `CMS_get0_SignerInfos()` could be used to check for content type correctness.
> 
> Unfortunately one issue remains, one cannot disambiguate between the CRLs field being absent or being empty. In some X.509/CRL profiles the presence of the `crls` field is illegal if it contains zero CRLs.

It should not and cannot be the job of the caller to try and find out in advance all error conditions that the function execution may suffer from. In particular since the internals of the function is allowed to change over time without notice.
This would work out only in case all its possible error conditions were documented - 
which is not feasible and would also put needless burden on all callers.

> Ergo, the `CMS_get1_crls()` interface was not great to begin with,

That's right.

> this commit only marginally made it better.

I disagree - the new semantics is the best one can do here without changing the signature of the function.
The only problem is that it is not fully backward compatible.

DDvO commented:
> Also of note, `CMS_get0_signers()` returns `NULL` on zero signers, not an empty `STACK` -
> so this change to some degree further distorts internal consistency.

In order to answer this, it should be clarified first whether an empty set of signers is to be considered an error or not.
* If it is an error, everything is fine and your "internal consistency" point does not apply.
* If it is not an error, then indeed its behavior is not consistent with the new one of `CMS_get1_{certs,crls}()`
  and its documentation does not correctly describe its suboptimal behavior in this case:
  > CMS_get0_signers() returns all signers or NULL if an error occurred.

t8m commented:
> * If it is an error, everything is fine with it and your "internal consistency" point does not apply.
> 
> * If it is not an error, then indeed its behavior is not consistent with the new one of `CMS_get1_{certs,crls}()`
>   and its documentation does not correctly describe its suboptimal behavior in this case:
>   > CMS_get0_signers() returns all signers or NULL if an error occurred.

It is somewhat in between - it would be an error to have no signers on fully created signed message. However if you are creating a message, CMS_get0_signers() can return NULL until you add signers.
